api/api/routers/mcp/_mcp_observability_middleware.py

```diff
@@ -0,0 +1,341 @@
+        """Create the callback handler for when tool call streaming completes."""
+
+        def on_streaming_complete(response_body: bytes) -> None:
+            """Process the complete response body after streaming is done"""
+
+            # Process tool call response
+            duration = time.time() - start_time
+            completed_at = datetime.now(timezone.utc)
+            tool_result = self._parse_tool_call_response(response_body, request_id)
+
+            # Create tool call data
+            tool_call_data = ToolCallData(
+                tool_name=tool_name,
+                tool_arguments=tool_arguments,
+                request_id=str(request_id),
+                duration=duration,
+                result=tool_result,
+                started_at=_tool_call_start_times.get(request_id, tool_call_start_time),
+                completed_at=completed_at,
+                user_agent=user_agent,
+            )
+
+            # Update session state with completed tool call (async call in sync context)
+            async def update_session():
+                await session_state.register_tool_call(tool_call_data)
+
+            # Clean up start time tracking
+            _tool_call_start_times.pop(request_id, None)
+
+            # Run observer agent in background
+            if response_body:
+                observer_data = ObserverAgentData(
+                    tool_name=tool_name,
+                    previous_tool_calls=session_state.tool_calls,
+                    tool_arguments=tool_arguments,
+                    tool_result=tool_result,
+                    duration_seconds=duration,
+                    user_agent=user_agent,
+                    mcp_session_id=session_state.session_id,
+                    request_id=str(request_id),
+                    organization_name=tenant_info.slug,
+                )
+
+                # Create background task for observer agent
+                async def run_observer_background():
+                    # Update session first, then run observer
+                    await update_session()
+                    await self._run_observer_agent_background(observer_data)
+
+                add_background_task(run_observer_background())
+            else:
+                # Just update session if no response body
+                add_background_task(update_session())
+                logger.warning(
+                    "No response body received, skipping observer agent",
+                    extra={
+                        "session_id": session_state.session_id,
+                        "request_id": request_id,
+                    },
+                )
+
+        return on_streaming_complete
+
+    async def dispatch(self, request: Request, call_next: Callable[[Request], Any]) -> Response:
+        """Main dispatch method with comprehensive error handling"""
+        # Setup session and validate request
+        setup_result = await self._setup_session_and_validate_request(request)
```

Comment:
can you add a TODO to remove the duplicate database calls ? We are validating the token and fetching the tenant twice, here and in the actual tool functions